README.rst
Introduction
============

See the full documentation on `readthedocs <http://collectiveelasticsearch.readthedocs.io/>`_.

Install elasticsearch
---------------------

Less than 5 minutes:
    - Download & install Java
    - Download & install Elastic Search
    - bin/elasticsearch

    Step by Step for Ubuntu:
        - add-apt-repository ppa:webupd8team/java
        - apt-get update
        - apt-get install git curl oracle-java7-installer
        - wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.7.0-linux-x86_64.tar.gz
        - wget https://artifacts.elastic.co/downloads/elasticsearch/elasticsearch-7.7.0-linux-x86_64.tar.gz.sha512
        - shasum -a 512 -c elasticsearch-7.7.0-linux-x86_64.tar.gz.sha512 
        - tar -xzf elasticsearch-7.7.0-linux-x86_64.tar.gz
        - cd elasticsearch
        - bin/elasticsearch &

Does it work?
    - curl http://localhost:9200/
    - Do you see the Hudsucker Proxy reference? "You Know, for Search"

Use Elastic Search in Plone:
    - Add collective.elasticsearch to eggs & re-run buildout
    - Restart Plone
    - Goto Control Panel
    - Add "Elastic Search" in Add-on Products
    - Click "Elastic Search" in "Add-on Configuration"
    - Enable
    - Click "Convert Catalog"
    - Click "Rebuild Catalog"

You now have a insanely scalable modern search engine. Now live the life of the Mind!

Overview
--------

This package aims to index all fields the portal_catalog indexes
and allows you to delete the `Title`, `Description` and `SearchableText`
indexes which can provide significant improvement to performance and RAM usage.

Then, ElasticSearch queries are ONLY used when Title, Description and SearchableText
text are in the query. Otherwise, Plone's default catalog will be used.
This is because Plone's default catalog is faster on normal queries than using
ElasticSearch.


Compatibility
-------------

Only tested with Plone 5 with Dexterity types.

It should also work with Plone 4.3 and archetypes.

Deployed with Elasticsearch 7.7.0

State
-----

Support for all index column types is done EXCEPT for the DateRecurringIndex
index column type. If you are doing a full text search along with a query that
contains a DateRecurringIndex column, it will not work.


Celery support
--------------

This package comes with Celery support where all indexing operations will be pushed
into celery to be run asynchronously.

Please see instructions for collective.celery to see how this works.


Running tests
-------------

Run elasticsearch for tests to utilize:

    docker run \
        -e "discovery.type=single-node" \
        -e "cluster.name=docker-cluster" \
        -e "ES_JAVA_OPTS=-Xms512m -Xmx512m" \
        -p 9200:9200 \
        docker.elastic.co/elasticsearch/elasticsearch-oss:7.7.0

Then, you can use one of the travis buildout test files:

    python bootstrap.py -c travis-5.0.cfg
    bin/buildout -c travis-5.0.cfg
    bin/test -s collective.elasticsearch


Travis
------

.. image:: https://travis-ci.org/collective/collective.elasticsearch.png
   :target: https://travis-ci.org/collective/collective.elasticsearch
   :alt: Travis CI


CHANGES.rst
Changelog
=========

3.0.5 (unreleased)
------------------

- `Issue#76 <https://github.com/collective/collective.elasticsearch/issues/76>`_ In 5.1+ we want to patch _unindexObject not unindexObject [ewohnlich]

- Explicit error logging added, if ES bulk action for indexing failed.
  [nazrulworld]

- Fix commit hook bug when content has been moved
  [instification]

- Add profile unistall
  [flavico]

- Adapted to Elastic Search 7.7
  [flavico]

- Possibility to apply analyzed index on mapping
  [flavico]


3.0.4 (2019-08-21)
------------------

- `Issue#63 <https://github.com/collective/collective.elasticsearch/issues/63>`_ Now ensuring unicode value would for both python2 and python3 case. [nazrulworld]

- Now possible to search by other than `Title`, `Description` and `SearchableText` indexes. [nazrulworld]


3.0.3 (2019-03-12)
------------------

- Add missing import logger
  [nazrulworld]


3.0.2 (2019-01-31)
------------------

- Fix Zope DateTime convert to also handle the datetime.date type [ewohnlich]


3.0.1 (2019-01-28)
------------------

- Fix sortable_title search issue
  [ewohnlich]


3.0.0 (2019-01-28)
------------------

- Fix date queries to work with `min:max` as well as `minmax`
  [vangheem]

- Fix sort order parsing and implementation
  [vangheem]

- Handle upgrades with missing `es_only_indexes` properly
  [vangheem]

- Add IReindexActive to request as a flag for other code
  [lucid-0]


2.0.2 (2018-11-27)
------------------


- Python 3 Support
  [vangheem]

- Support ES 6
  [lucid-0]

- Fix error causing "Server Status" on @@elastic-controlpanel to be empty.
  [fulv]


2.0.1 (2018-01-05)
------------------

- Prevent critical error when by chance query value is None.
  [thomasdesvenain]

- Minor code cleanup: readability, pep8, 80 cols, zca decorators.
  [jensens]

- Fix date criteria: 'minmax' instead of 'min:max' + string to date conversion
  [ebrehault]


2.0.0a6 (2017-03-29)
--------------------

- Gracefully handle upgrades in the settings interface so it doesn't break
  for people upgrading.
  [vangheem]


2.0.0a5 (2017-03-29)
--------------------

- Running indexing as admin as it is possible to initiate reindex or index on an
  object that you do not have permissions for
  [vangheem]


2.0.0a4 (2017-03-27)
--------------------

- released


2.0.0a3 (2017-03-27)
--------------------

- Add a method to set the body of the request during index creation.
  [Gagaro]

- Fixed get brain in lazy list with negative indexes.
  [thomasdesvenain]

- The list of indexes that forces es search is configurable.
  [thomasdesvenain]

- Works under Plone 4.3.
  [thomasdesvenain]

- Works with archetypes contents.
  [thomasdesvenain]

2.0.0a2 (2016-07-19)
--------------------

- We can pass a custom results factory and custom query parameters
  to IElasticSearchCatalog.search() method.
  So we can use it as a public interface for custom needs.
  [thomasdesvenain]

- Prevent from unindex before reindex when uid is unchanged, for instance at rename.
  Use a set for to-remove list.
  [thomasdesvenain]

- Fix indexing when removing the Title and Description indexes from Plone
  [vangheem]

2.0.0a1 (2016-06-06)
--------------------

- upgrade to elasticsearch 2.x
  [vangheem]

1.0.1a4 (2016-05-22)
--------------------

- provide better search query
  [vangheem]

1.0.1a3 (2016-03-22)
--------------------

- make sure to get alias definition right
  [vangheem]

1.0.1a2 (2016-03-18)
--------------------

- create index as an alias so you can potentially work on an existing alias without needing
  downtime
  [vangheem]

1.0.1a1 (2016-02-25)
--------------------

- change default sorting to descending.
  Closes: https://github.com/collective/collective.elasticsearch/issues/12
  [neilferreira]

1.0.0a1 (2016-02-25)
--------------------

- Initial release


src/collective/elasticsearch/tests/test_search.py
# -*- coding: utf-8 -*-
from collective.elasticsearch.testing import createObject
from collective.elasticsearch.testing import HAS_ATCONTENTTYPES
from collective.elasticsearch.tests import BaseFunctionalTest
from DateTime import DateTime

import time
import unittest2 as unittest


EVENT_KLASS = 'plone.app.event.dx.interfaces.IDXEvent'
DOCUMENT_KLASS = 'plone.app.contenttypes.interfaces.IDocument'


class TestQueries(BaseFunctionalTest):

    event_klass = EVENT_KLASS
    document_klass = DOCUMENT_KLASS

    def test_field_index_query(self):
        createObject(self.portal, 'Event', 'event', title='Some Event')
        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)
        el_results = self.catalog(portal_type='Event', Title='some event')
        self.assertEqual(len(el_results), 1)

    def test_keyword_index_query(self):
        createObject(self.portal, 'Event', 'event', title='Some Event')
        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)
        el_results = self.catalog(
            object_provides=[self.event_klass],
            SearchableText='Event')
        self.assertEqual(len(el_results), 1)

    def test_multi_keyword_index_query(self):
        createObject(self.portal, 'Event', 'event', title='New Content')
        createObject(self.portal, 'Document', 'page', title='New Content')
        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)
        el_results = self.catalog(
            object_provides=[self.event_klass, self.document_klass],
            SearchableText='new content')
        self.assertEqual(len(el_results), 2)

    def test_date_index_query(self):
        start = DateTime()
        time.sleep(1)
        events = []
        for idx in range(5):
            event = createObject(
                self.portal, 'Event',
                'event%i' % idx, title='Some Event %i' % idx,
                effective=DateTime('2015/09/25 20:00'))
            events.append(event)

        self.commit()
        self.es.connection.indices.flush()

        end = DateTime()
        query = {
            'query': (start, end),
            'range': 'minmax',
            'SearchableText': 'Event'
        }
        cat_results = self.catalog._old_searchResults(created=query)
        el_results = self.catalog(created=query)
        self.assertEqual(len(cat_results), len(el_results))
        self.assertEqual(len(cat_results), len(events))

        query = {'query': DateTime().latestTime(), 'range': 'min'}
        cat_results = self.catalog._old_searchResults(effective=query)
        el_results = self.catalog(effective=query)
        self.assertEqual(len(cat_results), len(el_results))
        self.assertEqual(len(cat_results), 0)

        query = {'query': DateTime().latestTime(), 'range': 'max'}
        cat_results = self.catalog._old_searchResults(effective=query)
        el_results = self.catalog(effective=query)
        self.assertEqual(len(cat_results), len(el_results))
        self.assertEqual(len(cat_results), 5)

    def test_text_index_query(self):
        for idx in range(5):
            createObject(
                self.portal, 'Document',
                'page%i' % idx, title='Page %i' % idx)
            # should not show up in results
        events = []
        for idx in range(5):
            event = createObject(
                self.portal, 'Event',
                'event%i' % idx, title='Some Event %i' % idx)
            events.append(event)

        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)

        el_results = self.catalog(Title='Some Event')
        self.assertEqual(len(el_results), len(events))

        el_results = self.catalog(Title='Some Event 1',
                                  sort_on='getObjPositionInParent')
        self.assertTrue('Some Event 1' in [
            b.Title for b in el_results])
        self.assertEqual(el_results[0].Title, 'Some Event 1')

    def test_path_index_query(self):
        folder1 = createObject(self.portal, 'Folder', 'folder0',
                               title='New Content 0')
        createObject(folder1, 'Document', 'page1', title='New Content 1')
        createObject(folder1, 'Document', 'page2', title='New Content 2')
        createObject(folder1, 'Document', 'page3', title='New Content 3')
        folder2 = createObject(folder1, 'Folder', 'folder4',
                               title='New Content 4')
        folder3 = createObject(folder2, 'Folder', 'folder5',
                               title='New Content 5')
        createObject(folder3, 'Document', 'page6', title='New Content 6')
        createObject(folder3, 'Document', 'page7', title='New Content 7')
        createObject(folder3, 'Document', 'page8', title='New Content 8')

        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)

        self.assertEqual(
            len(self.catalog(path={'depth': 0, 'query': '/plone/folder0'},
                             SearchableText='new content')), 1)
        self.assertEqual(
            len(self.catalog(path={'depth': 1, 'query': '/plone/folder0'},
                             SearchableText='new content')), 4)
        self.assertEqual(
            len(self.catalog(path={'depth': -1,
                                   'query': '/plone/folder0'},
                             SearchableText='new content')), 9)
        self.assertEqual(
            len(self.catalog(path={'depth': 1, 'query': '/plone'},
                             SearchableText='new content')), 1)
        # this proofs its wrong
        self.assertEqual(
            len(self.catalog(path={'query': '/plone/folder0',
                                   'navtree_start': 0, 'navtree': 1},
                             is_default_page=False,
                             SearchableText='new content')), 9)

    def test_combined_query(self):
        createObject(self.portal, 'Folder', 'folder1', title='Folder 1')
        self.commit()
        self.es.connection.indices.flush()
        self.assertEqual(
            len(self.catalog(path={'depth': 1, 'query': '/plone'},
                             portal_type='Folder',
                             is_default_page=False,
                             SearchableText='folder')),
            1)

    def test_brains(self):
        event = createObject(self.portal, 'Event', 'event', title='Some Event')
        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)
        el_results = self.catalog(portal_type='Event', Title='Some Event')
        self.assertEqual(len(el_results), 1)
        brain = el_results[0]
        self.assertEqual(brain.getObject(), event)
        self.assertEqual(brain.portal_type, 'Event')
        self.assertEqual(brain.getURL(), 'http://nohost/plone/event')
        self.assertEqual(brain.getPath(), '/plone/event')

        brain = el_results[-1]
        self.assertEqual(brain.getObject(), event)
        self.assertEqual(brain.portal_type, 'Event')
        self.assertEqual(brain.getURL(), 'http://nohost/plone/event')
        self.assertEqual(brain.getPath(), '/plone/event')

        createObject(self.portal, 'Event', 'event2', title='Some Event')
        self.commit()
        self.es.connection.indices.flush()
        time.sleep(1)

        el_results2 = self.catalog(
            portal_type='Event',
            Title='Some Event',
            sort_on='getId',
            sort_order='descending')
        self.assertEqual(len(el_results2), 2)
        brain = el_results2[0]
        self.assertEqual(brain.getId, 'event2')
        brain = el_results2[1]
        self.assertEqual(brain.getId, 'event')

        brain = el_results2[-1]
        self.assertEqual(brain.getId, 'event')
        brain = el_results2[-2]
        self.assertEqual(brain.getId, 'event2')

    def test_brains_indexing(self):
        for idx in range(120):
            createObject(
                self.portal, 'Document', '{0:04d}page'.format(idx),
                title='Page {}'.format(idx))
        self.commit()
        el_results = self.catalog(
            portal_type='Document',
            sort_on='getId',
            sort_order='asc')
        self.assertEqual(len(el_results), 120)
        self.assertEqual(el_results[0].getId, '0000page')
        self.assertEqual(el_results[-1].getId, '0119page')
        self.assertEqual(el_results[-50].getId, '0070page')
        self.assertEqual(el_results[-55].getId, '0065page')
        self.assertEqual(el_results[-100].getId, '0020page')


if HAS_ATCONTENTTYPES:
    from collective.elasticsearch.testing import ElasticSearch_FUNCTIONAL_TESTING_AT  # noqa

    EVENT_KLASS_AT = 'Products.ATContentTypes.interfaces.event.IATEvent'
    DOCUMENT_KLASS_AT = 'Products.ATContentTypes.interfaces.document.IATDocument'  # noqa

    class TestQueriesAT(TestQueries):
        layer = ElasticSearch_FUNCTIONAL_TESTING_AT

        event_klass = EVENT_KLASS_AT
        document_klass = DOCUMENT_KLASS_AT


def test_suite():
    return unittest.defaultTestLoader.loadTestsFromName(__name__)


src/collective/elasticsearch/es.py
# -*- coding: utf-8 -*-
from collective.elasticsearch import hook
from collective.elasticsearch import logger
from collective.elasticsearch.brain import BrainFactory
from collective.elasticsearch.interfaces import IElasticSearchCatalog
from collective.elasticsearch.interfaces import IElasticSettings
from collective.elasticsearch.interfaces import IMappingProvider
from collective.elasticsearch.interfaces import IQueryAssembler
from collective.elasticsearch.interfaces import IReindexActive
from collective.elasticsearch.utils import getESOnlyIndexes
from DateTime import DateTime
from elasticsearch import Elasticsearch
from elasticsearch.exceptions import NotFoundError, TransportError
from plone import api
from plone.registry.interfaces import IRegistry
from Products.CMFCore.permissions import AccessInactivePortalContent
from Products.CMFCore.utils import _checkPermission
from Products.CMFCore.utils import _getAuthenticatedUser
from Products.ZCatalog.Lazy import LazyMap
from zope.component import ComponentLookupError
from zope.component import getMultiAdapter
from zope.component import getUtility
from zope.globalrequest import getRequest
from zope.interface import implementer
from zope.interface import alsoProvides
import math

CONVERTED_ATTR = '_elasticconverted'
CUSTOM_INDEX_NAME_ATTR = '_elasticcustomindex'
INDEX_VERSION_ATTR = '_elasticindexversion'


class ElasticResult(object):

    def __init__(self, es, query, **query_params):
        assert 'sort' not in query_params
        assert 'start' not in query_params
        self.es = es
        self.bulk_size = es.get_setting('bulk_size', 50)
        qassembler = getMultiAdapter((getRequest(), es), IQueryAssembler)
        dquery, self.sort = qassembler.normalize(query)
        self.query = qassembler(dquery)

        # results are stored in a dictionary, keyed
        # but the start index of the bulk size for the
        # results it holds. This way we can skip around
        # for result data in a result object
        result = es._search(self.query, sort=self.sort, **query_params)['hits']
        self.results = {
            0: result['hits']
        }
        self.count = result['total']['value']
        self.query_params = query_params

    def __len__(self):
        return self.count

    def __getitem__(self, key):
        '''
        Lazy loading es results with negative index support.
        We store the results in buckets of what the bulk size is.
        This is so you can skip around in the indexes without needing
        to load all the data.
        Example(all zero based indexing here remember):
            (525 results with bulk size 50)
            - self[0]: 0 bucket, 0 item
            - self[10]: 0 bucket, 10 item
            - self[50]: 50 bucket: 0 item
            - self[55]: 50 bucket: 5 item
            - self[352]: 350 bucket: 2 item
            - self[-1]: 500 bucket: 24 item
            - self[-2]: 500 bucket: 23 item
            - self[-55]: 450 bucket: 19 item
        '''
        if isinstance(key, slice):
            return [self[i] for i in range(key.start, key.end)]
        else:
            if key + 1 > self.count:
                raise IndexError
            elif key < 0 and abs(key) > self.count:
                raise IndexError

            if key >= 0:
                result_key = int(key / self.bulk_size) * self.bulk_size
                start = result_key
                result_index = key % self.bulk_size
            elif key < 0:
                last_key = int(math.floor(
                    float(self.count) / float(self.bulk_size)
                )) * self.bulk_size
                start = result_key = last_key - (
                    (abs(key) / self.bulk_size) * self.bulk_size)
                if last_key == result_key:
                    result_index = key
                else:
                    result_index = (key % self.bulk_size) - (
                        self.bulk_size - (self.count % last_key)
                    )

            if result_key not in self.results:
                self.results[result_key] = self.es._search(
                    self.query, sort=self.sort, start=start,
                    **self.query_params)['hits']['hits']

            return self.results[result_key][result_index]


@implementer(IElasticSearchCatalog)
class ElasticSearchCatalog(object):
    '''
    from patched methods
    '''

    def __init__(self, catalogtool):
        self.catalogtool = catalogtool
        self.catalog = catalogtool._catalog

        try:
            registry = getUtility(IRegistry)
            try:
                self.registry = registry.forInterface(
                    IElasticSettings,
                    check=False
                )
            except Exception:
                self.registry = None
        except ComponentLookupError:
            self.registry = None

        self._conn = None

    @property
    def connection(self):
        if self._conn is None:
            kwargs = dict()
            if self.get_setting('timeout', 0):
                kwargs['timeout'] = self.get_setting('timeout')
            if self.get_setting('sniff_on_start', False):
                kwargs['sniff_on_start'] = True
            if self.get_setting('sniff_on_connection', False):
                kwargs['sniff_on_connection'] = True
            if self.get_setting('sniffer_timeout', 0):
                kwargs['sniffer_timeout'] = self.get_setting('sniffer_timeout')
            if self.get_setting('retry_on_timeout', False):
                kwargs['retry_on_timeout'] = True
            self._conn = Elasticsearch(
                self.registry.hosts,
                **kwargs
            )
        return self._conn

    def _search(self, query, sort=None, **query_params):
        '''
        '''
        if 'start' in query_params:
            query_params['from_'] = query_params.pop('start')

        query_params['stored_fields'] = query_params.get(
            'stored_fields', 'path.path')
        query_params['size'] = self.get_setting('bulk_size', 50)

        body = {'query': query}
        if sort is not None:
            body['sort'] = sort

        return self.connection.search(index=self.index_name,
                                      body=body,
                                      **query_params)

    def search(self, query, factory=None, **query_params):
        """
        @param query: dict
            The plone query
        @param factory: function(result: dict): any
            The factory that maps each elastic search result.
            By default, get the plone catalog brain.
        @param query_params:
            Parameters to pass to the search method
            'stored_fields': the list of fields to get from stored source
        @return: LazyMap
        """
        result = ElasticResult(self, query, **query_params)
        if not factory:
            factory = BrainFactory(self.catalog)
        return LazyMap(factory, result, result.count)

    @property
    def catalog_converted(self):
        return getattr(self.catalogtool, CONVERTED_ATTR, False)

    @property
    def enabled(self):
        return (self.registry
                and self.registry.enabled
                and self.catalog_converted
                )

    def get_setting(self, name, default=None):
        return getattr(self.registry, name, default)

    def catalog_object(self, obj, uid=None, idxs=[], update_metadata=1,
                       pghandler=None):
        if idxs != ['getObjPositionInParent']:
            self.catalogtool._old_catalog_object(
                obj, uid, idxs, update_metadata, pghandler)

        if not self.enabled:
            return
        hook.add_object(self, obj)

    def uncatalog_object(self, uid, obj=None, *args, **kwargs):
        # always need to uncatalog to remove brains, etc
        if obj is None:
            # with archetypes, the obj is not passed, only the uid is
            try:
                obj = api.content.get(uid)
            except KeyError:
                pass

        result = self.catalogtool._old_uncatalog_object(uid, *args, **kwargs)
        if self.enabled:
            hook.remove_object(self, obj)

        return result

    def manage_catalogRebuild(self, *args, **kwargs):
        if self.registry.enabled:
            self.recreateCatalog()

        alsoProvides(getRequest(), IReindexActive)
        return self.catalogtool._old_manage_catalogRebuild(*args, **kwargs)

    def manage_catalogClear(self, *args, **kwargs):
        if self.enabled:
            self.recreateCatalog()

        return self.catalogtool._old_manage_catalogClear(*args, **kwargs)

    def recreateCatalog(self):
        conn = self.connection

        try:
            conn.indices.delete(index=self.real_index_name)
        except NotFoundError:
            pass
        except TransportError as exc:
            if exc.error != 'illegal_argument_exception':
                raise
            conn.indices.delete_alias(index="_all", name=self.real_index_name)

        if self.index_version:
            try:
                conn.indices.delete_alias(
                    self.index_name,
                    self.real_index_name)
            except NotFoundError:
                pass
        self.convertToElastic()

    def searchResults(self, REQUEST=None, check_perms=False, **kw):
        enabled = False
        if self.enabled:
            # need to also check if it is a search result we care about
            # using EL for
            if getESOnlyIndexes().intersection(kw.keys()):
                enabled = True
        if not enabled:
            if check_perms:
                return self.catalogtool._old_searchResults(REQUEST, **kw)
            else:
                return self.catalogtool._old_unrestrictedSearchResults(
                    REQUEST,
                    **kw)

        if isinstance(REQUEST, dict):
            query = REQUEST.copy()
        else:
            query = {}
        query.update(kw)

        if check_perms:
            show_inactive = query.get('show_inactive', False)
            if isinstance(REQUEST, dict) and not show_inactive:
                show_inactive = 'show_inactive' in REQUEST

            user = _getAuthenticatedUser(self.catalogtool)
            query['allowedRolesAndUsers'] = \
                self.catalogtool._listAllowedRolesAndUsers(user)

            if not show_inactive and not _checkPermission(
                    AccessInactivePortalContent, self.catalogtool):
                query['effectiveRange'] = DateTime()
        orig_query = query.copy()
        logger.debug('Running query: %s' % repr(orig_query))
        try:
            results = self.search(query)
            return results
        except Exception:
            logger.error(
                'Error running Query: {0!r}'.format(orig_query), exc_info=True)
            return self.catalogtool._old_searchResults(REQUEST, **kw)

    def convertToElastic(self):
        setattr(self.catalogtool, CONVERTED_ATTR, True)
        self.catalogtool._p_changed = True
        adapter = getMultiAdapter((getRequest(), self), IMappingProvider)
        mapping = adapter()
        self.connection.indices.put_mapping(
            body=mapping,
            index=self.index_name)

    @property
    def index_name(self):
        if hasattr(self.catalogtool, CUSTOM_INDEX_NAME_ATTR):
            return getattr(self.catalogtool, CUSTOM_INDEX_NAME_ATTR)
        return '-'.join(self.catalogtool.getPhysicalPath()[1:]).lower()

    @property
    def index_version(self):
        return getattr(self.catalogtool, INDEX_VERSION_ATTR, None)

    def bump_index_version(self):
        version = getattr(self.catalogtool, INDEX_VERSION_ATTR, None)
        if version is None:
            version = 1
        else:
            version += 1
        setattr(self.catalogtool, INDEX_VERSION_ATTR, version)
        self.catalogtool._p_changed = True
        return version

    @property
    def real_index_name(self):
        if self.index_version:
            return '%s_%i' % (self.index_name, self.index_version)
        return self.index_name

    @property
    def doc_type(self):
        return self.catalogtool.getId().lower()
